Gillian-C/lib/Chunk.ml

```diff
@@ -33,21 +53,55 @@ let to_string = function
   | Mint64 -> "int64"
   | Mfloat32 -> "float32"
   | Mfloat64 -> "float64"
-  | Many32 -> "any32"
-  | Many64 -> "any64"
+  | Mptr -> "ptr"
 
 let pp fmt chunk = Fmt.pf fmt "%s" (to_string chunk)
-let type_of = Compcert.AST.type_of_chunk
+
+let type_of = function
+  | Mint64 -> Compcert.AST.Tlong
+  | Mfloat32 -> Compcert.AST.Tsingle
+  | Mfloat64 -> Compcert.AST.Tfloat
+  | Mptr ->
+      if Compcert.Archi.ptr64 then Compcert.AST.Tany64 else Compcert.AST.Tany32
+  | _ -> Tint
 
 let size chunk =
   let open Compcert in
-  Camlcoq.Z.to_int (Memdata.size_chunk chunk)
+  let open BinNums in
+  let value =
+    match chunk with
+    | Mint8signed -> Memdata.size_chunk Mint8signed
+    | Mint8unsigned -> Memdata.size_chunk Mint8unsigned
+    | Mint16signed -> Memdata.size_chunk Mint16signed
+    | Mint16unsigned -> Memdata.size_chunk Mint16unsigned
+    | Mint32 -> Memdata.size_chunk Mint32
+    | Mint64 -> Memdata.size_chunk Mint64
+    | Mfloat32 -> Memdata.size_chunk Mfloat32
+    | Mfloat64 -> Memdata.size_chunk Mfloat64
+    | Mptr ->
+        if Compcert.Archi.ptr64 then Zpos (Coq_xO (Coq_xO (Coq_xO Coq_xH)))
+        else Zpos (Coq_xO (Coq_xO Coq_xH))
+  in
+  Camlcoq.Z.to_int value
 
 let size_expr chunk = Gil_syntax.Expr.int (size chunk)
 
 let align chunk =
   let open Compcert in
-  Camlcoq.Z.to_int (Memdata.align_chunk chunk)
+  let open BinNums in
+  let value =
+    match chunk with
+    | Mint8signed -> Memdata.align_chunk Mint8signed
+    | Mint8unsigned -> Memdata.align_chunk Mint8unsigned
+    | Mint16signed -> Memdata.align_chunk Mint16signed
+    | Mint16unsigned -> Memdata.align_chunk Mint16unsigned
+    | Mint32 -> Memdata.align_chunk Mint32
+    | Mint64 -> Memdata.align_chunk Mint64
+    | Mfloat32 -> Memdata.align_chunk Mfloat32
+    | Mfloat64 -> Memdata.align_chunk Mfloat64
+    | Mptr -> Zpos (Coq_xO (Coq_xO Coq_xH))
+  in
+  Camlcoq.Z.to_int value
 
-let equal = Compcert.AST.chunk_eq
-let ptr = Compcert.AST.coq_Mptr
+let equal chunk1 chunk2 = chunk1 = chunk2
```

Comment:
Please use `[@@deriving eq]` on `type t` instead.
Polymorphic equality is much slower than using an actual function that explains how to test for equality in ocaml